### PR TITLE
Server Module Update

### PR DESCRIPTION
**Pull request by kiroco12**
This is the last part to update the core. Server module has been updated to Scala 2.12.10 and it's dependencies. Changed free algebra to tagless final.

### CODE REVIEW

**Comment by fedefernandez** [comment]
@kiroco12 this looks to be a problem on one recursive call that is being executed in the akka thread. I'd say that is not related to akka but something in our code that is generating an endless recursive call, causing the `StackOverflowError` error. Sometimes can we due to a serialization of a large object.

File: project/ProjectPlugin.scala [hunk 1]
```diff
@@ -21,31 +21,38 @@ object ProjectPlugin extends AutoPlugin {
       Map(
         'evaluator     -> "0.5.0-SNAPSHOT",
```
Comment (juanpedromoreno) [inline project/ProjectPlugin.scala]: We don't need this key anymore

File: project/plugins.sbt [hunk 2]
```diff
@@ -11,10 +11,14 @@ resolvers ++= Seq(
...
 // Sbt plugins
-addSbtPlugin("org.scala-js"       % "sbt-scalajs"              % "0.6.29")
-addSbtPlugin("com.eed3si9n"       % "sbt-buildinfo"            % "0.9.0")
-addSbtPlugin("com.47deg"          % "sbt-org-policies"         % "0.12.0-M3")
-addSbtPlugin("com.typesafe.play"  % "sbt-plugin"               % "2.8.0-M5")
-addSbtPlugin("org.portable-scala" % "sbt-scalajs-crossproject" % "0.6.1")
-addSbtPlugin("com.vmunier"        % "sbt-web-scalajs"          % "1.0.9-0.6")
-addSbtPlugin("com.typesafe.sbt"   % "sbt-gzip"                 % "1.0.2")
+addSbtPlugin("org.irundaia.sbt"   % "sbt-sassify"               % "1.4.13")
+addSbtPlugin("org.scala-js"       % "sbt-scalajs"               % "0.6.29")
```
Comment (juanpedromoreno) [inline project/plugins.sbt]: We should bump to 0.6.31

File: project/ProjectPlugin.scala [hunk 3]
```diff
@@ -21,31 +21,38 @@ object ProjectPlugin extends AutoPlugin {
...
+        'github4s            -> "0.20.1",
+        'classutil           -> "1.5.1",
+        'commonsio           -> "2.6",
+        'highlightjs         -> "9.15.10",
+        'knockoff            -> "0.8.12",
+        'newrelic            -> "5.8.0",
+        'postgres            -> "42.2.8",
+        'scalajsscripts      -> "1.1.4",
+        'scalariform         -> "1.8.3",
+        'scalatest           -> "3.0.8",
```
Comment (juanpedromoreno) [inline project/ProjectPlugin.scala]: We should bump to 3.1.0. Keep in mind that the artifact has been deduplicated into two, where you might need `org.scalatestplus.scalatestplus-scalacheck`

> In 3.0.8 we moved Checkers under org.scalatest.prop to org.scalatestplus.scalacheck package, and deprecated Checkers under org.scalatest.prop. Checkers was still released in the ScalaTest artifact, with the org.scalatestplus.scalacheck package name. In 3.1.0, we are removing the deprecated name and the actual implementation in the new package from the ScalaTest artifact. From now on these will be released as a separate, optional module to allow ScalaTest and its ScalaCheck integration library to evolve more independently.

Comment (kiroco12) [reply project/ProjectPlugin.scala]: Should I update exercises too? The libraries are using 3.0.8

Comment (juanpedromoreno) [reply project/ProjectPlugin.scala]: Yes, we can do it later on

File: build.sbt [hunk 4]
```diff
@@ -101,6 +114,19 @@ lazy val client = (project in file("client"))
...
+lazy val `evaluator-client` = (project in file("eval-client"))
+  .enablePlugins(AutomateHeaderPlugin)
+  .settings(
+    name := "evaluator-client",
+    libraryDependencies ++= Seq(
+      "org.http4s"    %% "http4s-blaze-client" % v('http4s),
+      "org.http4s"    %% "http4s-circe"        % v('http4s),
+      "io.circe"      %% "circe-core"          % v('circeversion),
+      "io.circe"      %% "circe-generic"       % v('circeversion),
+      "org.scalatest" %% "scalatest"           % v('scalatest) % "test"
```
Comment (juanpedromoreno) [inline build.sbt]:
```suggestion
      "org.scalatest" %% "scalatest"           % v('scalatest) % Test
```

File: compiler/src/main/scala/org/scalaexercises/exercises/compiler/Compiler.scala [hunk 5]
```diff
@@ -235,8 +235,9 @@ case class Compiler() {
         )
     }
 
-    def oneline(msg: String) = {
-      val msg0 = msg.lines.mkString(s"${Console.BLUE}\\n${Console.RESET}")
+    // TODO: Not being used anywhere. Can this be removed?
+    /*def oneline(msg: String) = {
```
Comment (juanpedromoreno) [inline compiler/src/main/scala/org/scalaexercises/exercises/compiler/Compiler.scala]: If it's not needed, let's remove this commented function along with the TODO comment

File: compiler/src/main/scala/org/scalaexercises/exercises/compiler/Compiler.scala [hunk 6]
```diff
@@ -327,7 +328,7 @@ case class Compiler() {
         val libraryAsDependency =
           s"${buildMetaInfo.organization}:${buildMetaInfo.name}_${buildMetaInfo.scalaVersion
             .substring(0, 4)}:${buildMetaInfo.version}"
-        libraryAsDependency :: buildMetaInfo.libraryDependencies.toList
+        List(libraryAsDependency) // :: buildMetaInfo.libraryDependencies.toList
```
Comment (juanpedromoreno) [inline compiler/src/main/scala/org/scalaexercises/exercises/compiler/Compiler.scala]:
```suggestion
        List(libraryAsDependency) // :: buildMetaInfo.libraryDependencies.toList
        // Evaluator can resolve transitive dependencies, the reason of only passing the library as a dependency
```

File: compiler/src/main/scala/org/scalaexercises/exercises/compiler/comments.scala [hunk 7]
```diff
@@ -226,7 +226,7 @@ private[compiler] object CommentParsing {
       val name1 = params.get("name").collect {
         case Body(List(Paragraph(Chain(List(Summary(Text(value))))))) => value.trim
       }
-      val name2 = name1.filter(_.lines.length == 1)
+      val name2 = name1.filter(n => n.linesIterator.length == 1)
```
Comment (juanpedromoreno) [inline compiler/src/main/scala/org/scalaexercises/exercises/compiler/comments.scala]:
```suggestion
      val name2 = name1.filter(_.linesIterator.length == 1)
```

File: project/ProjectPlugin.scala [hunk 8]
```diff
@@ -19,33 +19,40 @@ object ProjectPlugin extends AutoPlugin {
...
+        'scalariform         -> "1.8.3",
+        'scalatest           -> "3.1.0",
+        'scalatestplusScheck -> "3.1.0.0-RC2",
+        'scalacheckversion   -> "1.14.2",
+        'scalacheckshapeless -> "1.2.3",
+        'upickle             -> "0.7.5",
+        'webjars             -> "2.7.3",
+        'scalamacros         -> "2.1.1",
+        'monix               -> "3.1.0",
+        'http4s              -> "0.20.13",
```
Comment (juanpedromoreno) [inline project/ProjectPlugin.scala]:
```suggestion
        'http4s              -> "0.20.15",
```

?

File: project/ProjectPlugin.scala [hunk 9]
```diff
@@ -93,10 +100,10 @@ object ProjectPlugin extends AutoPlugin {
         Resolver.mavenLocal,
         Resolver.sonatypeRepo("snapshots"),
         Resolver.sonatypeRepo("releases"),
-        Resolver.bintrayIvyRepo("sbt-plugins", "https://dl.bintray.com/ssidorenko/sbt-plugins/"),
-        Resolver.bintrayIvyRepo("scalaz-bintray", "http://dl.bintray.com/scalaz/releases"),
-        Resolver.typesafeIvyRepo("http://repo.typesafe.com/typesafe/releases/"),
-        Resolver.typesafeRepo("http://repo.typesafe.com/typesafe/maven-releases/")
+        Resolver.bintrayIvyRepo("ssidorenko", "sbt-plugins"),
+        //Resolver.bintrayIvyRepo("scalaz-bintray", "http://dl.bintray.com/scalaz/releases"),
```
Comment (juanpedromoreno) [inline project/ProjectPlugin.scala]: 🔥

File: project/ProjectPlugin.scala [hunk 10]
```diff
@@ -106,7 +112,8 @@ object ProjectPlugin extends AutoPlugin {
       headerLicense := Some(
         ScalaExercisesLicense("2015-2019", "47 Degrees, LLC. <http://www.47deg.com>")),
       ScoverageKeys.coverageFailOnMinimum := false
-    ) ++ addCompilerPlugin("org.scalamacros" % "paradise" % v('scalamacros) cross CrossVersion.full) ++ shellPromptSettings
+    ) ++ addCompilerPlugin("org.scalamacros" % "paradise"            % v('scalamacros) cross CrossVersion.full) ++
+      addCompilerPlugin("com.olegpy"         %% "better-monadic-for" % "0.3.1") ++ shellPromptSettings
```
Comment (juanpedromoreno) [inline project/ProjectPlugin.scala]: Does it make sense to move 0.3.1 to its own `val`?

File: runtime/src/main/scala/org/scalaexercises/exercises/Exercises.scala [hunk 11]
```diff
@@ -19,25 +19,32 @@
...
 
   private[this] def classMap(cl: ClassLoader) = {
-    val files       = cl.asInstanceOf[URLClassLoader].getURLs map (_.getFile)
-    val classFinder = ClassFinder(files map (new File(_)) filter (f ⇒ f.exists()))
-    val classes     = classFinder.getClasses.toIterator
-    ClassFinder.classInfoMap(classes)
+    val files = cl
+      .asInstanceOf[URLClassLoader]
+      .getURLs
+      .map(url => new File(url.getFile)) filter (f ⇒ f.exists())
```
Comment (juanpedromoreno) [inline runtime/src/main/scala/org/scalaexercises/exercises/Exercises.scala]:
```suggestion
      .map(url => new File(url.getFile)) filter (_.exists())
```

File: server/app/org/scalaexercises/exercises/ApplicationLoader.scala [hunk 12]
```diff
@@ -12,79 +14,117 @@
...
     with EvolutionsComponents
     with HikariCPComponents {
 
   applicationEvolutions.start()
 
-  override def dynamicEvolutions: DynamicEvolutions = new DynamicEvolutions
+  override lazy val dynamicEvolutions: DynamicEvolutions = new DynamicEvolutions
+
+  lazy val aa = (for {
+    ec <- ExecutionContexts.fixedThreadPool[IO](32)
```
Comment (juanpedromoreno) [inline server/app/org/scalaexercises/exercises/ApplicationLoader.scala]: Let's move the 32 magic number to its configuration setting

File: server/app/org/scalaexercises/exercises/ApplicationLoader.scala [hunk 13]
```diff
@@ -12,79 +14,117 @@
...
+    with CaffeineCacheComponents
     with EvolutionsComponents
     with HikariCPComponents {
 
   applicationEvolutions.start()
 
-  override def dynamicEvolutions: DynamicEvolutions = new DynamicEvolutions
+  override lazy val dynamicEvolutions: DynamicEvolutions = new DynamicEvolutions
+
+  lazy val aa = (for {
```
Comment (juanpedromoreno) [inline server/app/org/scalaexercises/exercises/ApplicationLoader.scala]: Could we rename `aa` to something more meaningful?

File: server/app/org/scalaexercises/exercises/controllers/ApplicationController.scala [hunk 14]
```diff
@@ -12,131 +14,126 @@
...
-          case (libraries, user, Some(token), progress, _) ⇒
+          case (libraries, user, Some(_), progress, _) ⇒
             Ok(
               views.html.templates.home
-                .index(user = user, libraries = libraries, progress = progress, repo = repo))
+                .index(
+                  user = user,
+                  libraries = libraries,
+                  progress = progress,
+                  repo = repo.getOrElse(Repository(0, 0, 0)),
```
Comment (juanpedromoreno) [inline server/app/org/scalaexercises/exercises/controllers/ApplicationController.scala]: 0, 0, 0? what is that?

File: server/app/org/scalaexercises/exercises/controllers/ExercisesController.scala [hunk 15]
```diff
@@ -12,86 +14,79 @@
...
+    userOps: UserOps[IO],
+    userProgressOps: UserProgressOps[IO],
+    mode: Mode)
+    extends BaseController
     with JsonFormats
-    with AuthenticationModule
-    with ProdInterpreters {
+    with AuthenticationModule {
+
+  private implicit val ce: ConcurrentEffect[IO] =
```
Comment (juanpedromoreno) [inline server/app/org/scalaexercises/exercises/controllers/ExercisesController.scala]: These kind of implicits can be provided implicitly

File: server/app/org/scalaexercises/exercises/controllers/ApplicationController.scala (hunk 14 above)
Comment (kiroco12) [reply server/app/org/scalaexercises/exercises/controllers/ApplicationController.scala]: Is a default value for the numbers of Scala Exercises repo shown at the index if connection with GitHub fails. Any alternative?

File: sbt-exercise/src/sbt-test/sbt-exercise/basic/project/build.properties [hunk 16]
```diff
@@ -0,0 +1 @@
+sbt.version=1.2.8
```
Comment (juanpedromoreno) [inline sbt-exercise/src/sbt-test/sbt-exercise/basic/project/build.properties]: In the scripted  tests, the `build.properties` properties is not usually included to be aligned with the outer sbt version

File: sbt-exercise/src/sbt-test/sbt-exercise/scalariform/project/build.properties [hunk 17]
```diff
@@ -0,0 +1 @@
+sbt.version=1.2.8
```
Comment (juanpedromoreno) [inline sbt-exercise/src/sbt-test/sbt-exercise/scalariform/project/build.properties]: Same here, let's remove all the build.properties files within the scripted tests

File: server/app/org/scalaexercises/exercises/controllers/ExercisesController.scala [hunk 18]
```diff
@@ -126,22 +117,19 @@ class ExercisesController(
         def logError(method: String, mainMsg: String, ex: Option[Throwable] = None) = {
           val msg = s"[$method] $mainMsg. ExerciseEvaluation: $evaluation with user $user"
           ex match {
-            case Some(e) ⇒ Logger.error(msg, e)
-            case None    ⇒ Logger.error(msg)
+            case Some(e) ⇒ Logger(this.getClass).error(msg, e)
+            case None    ⇒ Logger(this.getClass).error(msg)
```
Comment (juanpedromoreno) [inline server/app/org/scalaexercises/exercises/controllers/ExercisesController.scala]: We could initialize the logger as part of the class and we are not creating new instances every time we are using it

File: server/app/org/scalaexercises/exercises/controllers/ApplicationController.scala [hunk 19]
```diff
@@ -12,131 +14,126 @@
...
+    with AuthenticationModule {
+  lazy val topLibraries: List[String] = config.getOptional[Seq[String]]("exercises.top_libraries") map (_.toList) getOrElse Nil
 
   val MainRepoCacheKey = "scala-exercises.repo"
 
+  val configUtils = ConfigUtils(config)
+
+  private val service = new ExercisesService(cl)
+
+  private implicit val cs = IO.contextShift(defaultExecutionContext)
```
Comment (juanpedromoreno) [inline server/app/org/scalaexercises/exercises/controllers/ApplicationController.scala]: This comment applies to the entire application.

Could we pass this kind of instances (ExecutionContexts, ContextShifts, Timers, etc.) implicitly to every controller and service? So we are building just one instance of every time once, and only at the beginning (when the application starts).

Comment (kiroco12) [reply server/app/org/scalaexercises/exercises/controllers/ApplicationController.scala]: It has sense when we're talking about configuration, application mode... to be a val. Should I declare the rest (ExecutionContexts, Timers...) as a def instead val?

File: server/app/org/scalaexercises/exercises/ApplicationLoader.scala [hunk 20]
```diff
@@ -12,79 +14,130 @@
...
-  val exercisesController    = new ExercisesController
-  val userController         = new UserController
-  val oauthController        = new OAuthController
-  val userProgressController = new UserProgressController
-  val loaderIOController     = new LoaderIOController
-  val sitemapController      = new SitemapController
+  implicit def cs: ContextShift[IO] = IO.contextShift(controllerComponents.executionContext)
 
-  val assets = new _root_.controllers.Assets(httpErrorHandler)
+  implicit def ce: ConcurrentEffect[IO] = IO.ioConcurrentEffect(cs)
```
Comment (juanpedromoreno) [inline server/app/org/scalaexercises/exercises/ApplicationLoader.scala]: Could be this a `val`?

Comment (juanpedromoreno) [inline server/app/org/scalaexercises/exercises/ApplicationLoader.scala]:
```suggestion
  implicit lazy val ce: ConcurrentEffect[IO] = IO.ioConcurrentEffect(cs)
```

File: server/app/org/scalaexercises/exercises/ApplicationLoader.scala [hunk 21]
```diff
@@ -12,79 +14,130 @@
...
+  implicit val mode = environment.mode
 
-  val applicationController  = new ApplicationController(defaultCacheApi)
-  val exercisesController    = new ExercisesController
-  val userController         = new UserController
-  val oauthController        = new OAuthController
-  val userProgressController = new UserProgressController
-  val loaderIOController     = new LoaderIOController
-  val sitemapController      = new SitemapController
+  implicit def cs: ContextShift[IO] = IO.contextShift(controllerComponents.executionContext)
```
Comment (juanpedromoreno) [inline server/app/org/scalaexercises/exercises/ApplicationLoader.scala]:
```suggestion
  implicit lazy val cs: ContextShift[IO] = IO.contextShift(controllerComponents.executionContext)
```

File: server/app/org/scalaexercises/exercises/services/handlers/ExerciseOpsHandler.scala [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+
+package org.scalaexercises.exercises.services.handlers
+
+import cats.effect.Sync
+import org.scalaexercises.algebra.exercises.ExerciseOps
+import org.scalaexercises.exercises.services.ExercisesService
+import org.scalaexercises.types.exercises.ExerciseEvaluation.EvaluationRequest
+import org.scalaexercises.types.exercises.{ExerciseEvaluation, Library, Section}
+
+class ExerciseOpsHandler[F[_]: Sync](implicit F: Sync[F], service: ExercisesService)
```
Comment (juanpedromoreno) [inline server/app/org/scalaexercises/exercises/services/handlers/ExerciseOpsHandler.scala]: Duplicated:

```suggestion
class ExerciseOpsHandler[F[_]](implicit F: Sync[F], service: ExercisesService)
```

**Pull request merged**